### PR TITLE
Do authorization checks for several PFS calls

### PR DESCRIPTION
**Pull request by msteffen**
Specifically: {Start,Finish,Inspect,List,Delete}Commit, {List,Set,Delete}Branch, and {Put,Get,Inspect,List,Glob,Diff,Delete}File

### CODE REVIEW

File: src/server/auth/server/auth.go [hunk 1]
```diff
@@ -281,14 +281,37 @@ func (a *apiServer) SetScope(ctx context.Context, req *authclient.SetScopeReques
 
 		var acl authclient.ACL
 		if err := acls.Get(req.Repo.Name, &acl); err != nil {
-			return fmt.Errorf("ACL not found for repo %v", req.Repo.Name)
+			// Might be creating a new ACL. Check that 'req' sets an owner
+			if req.Scope != authclient.Scope_OWNER {
+				return fmt.Errorf("ACL not found for repo %v", req.Repo.Name)
+			}
 		}
```
Comment (derekchiang) [inline src/server/auth/server/auth.go]: `acl.Entries` would be a nil map if the `Get` above failed, I think.

File: src/server/pfs/server/driver.go [hunk 2]
```diff
@@ -412,7 +428,26 @@ func (d *driver) buildCommit(ctx context.Context, parent *pfs.Commit, branch str
 	return d.makeCommit(ctx, parent, branch, provenance, tree)
 }
 
+// d.CheckIsAuthorized returns an error if the current user (in 'ctx') has
+// authorization scope 's' for repo 'r'
+func (d *driver) CheckIsAuthorized(ctx context.Context, r *pfs.Repo, s auth.Scope) error {
```
Comment (derekchiang) [inline src/server/pfs/server/driver.go]: Does this method need to be public?

File: src/server/pfs/server/driver.go [hunk 3]
```diff
@@ -621,6 +651,10 @@ func (d *driver) finishCommit(ctx context.Context, commit *pfs.Commit) error {
 // This side effect is used internally by other APIs to resolve branch
 // names to real commit IDs.
 func (d *driver) inspectCommit(ctx context.Context, commit *pfs.Commit) (*pfs.CommitInfo, error) {
+	d.initializePachConn()
+	if err := d.CheckIsAuthorized(ctx, commit.Repo, auth.Scope_READER); err != nil {
```
Comment (derekchiang) [inline src/server/pfs/server/driver.go]: This should probably be put below the nil commit check below.

File: src/server/pfs/server/driver.go [hunk 4]
```diff
@@ -869,6 +908,7 @@ func (d *driver) subscribeCommit(ctx context.Context, repo *pfs.Repo, branch str
 }
 
 func (d *driver) flushCommit(ctx context.Context, fromCommits []*pfs.Commit, toRepos []*pfs.Repo) (CommitStream, error) {
+	d.initializePachConn()
```
Comment (derekchiang) [inline src/server/pfs/server/driver.go]: Or here?

File: src/server/pfs/server/driver.go [hunk 5]
```diff
@@ -150,7 +159,10 @@ func newLocalDriver(blockAddress string, etcdPrefix string) (*driver, error) {
 	return newDriver(blockAddress, []string{"localhost:32379"}, etcdPrefix, defaultCacheSize)
 }
 
-func (d *driver) getObjectClient() (*client.APIClient, error) {
+// TODO(msteffen): client initialization (both etcd and pachd) might be better
+// placed happen in server.go, near main(), so that we only pay the dial cost
+// once, and so that pps doesn't need to have its own initialization code
+func (d *driver) initializePachConn() error {
```
Comment (derekchiang) [inline src/server/pfs/server/driver.go]: I wonder if it makes sense to put this into `CheckIsAuthorized` to reduce boilerplate.

File: src/server/auth/server/auth.go (hunk 1 above)
Comment (msteffen) [inline src/server/auth/server/auth.go]: Good catch, done

File: src/server/pfs/server/driver.go (hunk 5 above)
Comment (msteffen) [inline src/server/pfs/server/driver.go]: Done

File: src/server/pfs/server/driver.go (hunk 2 above)
Comment (msteffen) [inline src/server/pfs/server/driver.go]: Nope, done.

File: src/server/pfs/server/driver.go (hunk 4 above)
Comment (msteffen) [inline src/server/pfs/server/driver.go]: Done.

File: src/server/pfs/server/driver.go (hunk 3 above)
Comment (msteffen) [inline src/server/pfs/server/driver.go]: Done.

File: src/server/pfs/server/driver.go [hunk 6]
```diff
@@ -150,21 +162,41 @@ func newLocalDriver(blockAddress string, etcdPrefix string) (*driver, error) {
...
-		}
+		return d.onceErr
+	}
+	return nil
+}
+
+// checkIsAuthorized returns an error if the current user (in 'ctx') has
+// authorization scope 's' for repo 'r'
+func (d *driver) checkIsAuthorized(ctx context.Context, r *pfs.Repo, s auth.Scope) error {
+	initializePachConn()
```
Comment (derekchiang) [inline src/server/pfs/server/driver.go]: This wouldn't compile

File: src/server/pfs/server/driver.go [hunk 7]
```diff
@@ -413,6 +447,9 @@ func (d *driver) buildCommit(ctx context.Context, parent *pfs.Commit, branch str
 }
 
 func (d *driver) makeCommit(ctx context.Context, parent *pfs.Commit, branch string, provenance []*pfs.Commit, treeRef *pfs.Object) (*pfs.Commit, error) {
+	if err := d.CheckIsAuthorized(ctx, parent.Repo, auth.Scope_WRITER); err != nil {
```
Comment (derekchiang) [inline src/server/pfs/server/driver.go]: Need to change the function names.

**Pull request merged**